File: typescript/src/pruner.ts
/**
 * neuromem/pruner
 * ---------------
 * Safely prunes low-importance messages while respecting token budgets
 * and preserving critical content.
 */

import { Message, MessageScorer, ScoredMessage, estimateTokens } from "./scorer.js";
import { Summarizer, SummaryResult, summaryAsMessage } from "./summarizer.js";

// ---------------------------------------------------------------------------
// Result type
// ---------------------------------------------------------------------------

export interface PruneResult {
  messages: Message[];
  removedCount: number;
  removedTokens: number;
  summaryInserted: boolean;
  summaryResult?: SummaryResult;
  scores: ScoredMessage[];
}

// ---------------------------------------------------------------------------
// PrunerOptions
// ---------------------------------------------------------------------------

export interface PrunerOptions {
  /** Max total tokens. Default: 4096 */
  tokenBudget?: number;
  /** Messages below this score are candidates for pruning. Default: 0.3 */
  minScoreThreshold?: number;
  /** Always preserve the most recent N turns. Default: 4 */
  alwaysKeepLastN?: number;
  /** Try summarizing before hard-dropping. Default: true */
  summarizeBeforePrune?: boolean;
  scorer?: MessageScorer;
  summarizer?: Summarizer;
  /** Roles that must never be pruned. Default: ["system"] */
  preserveRoles?: string[];
}

// ---------------------------------------------------------------------------
// Pruner
// ---------------------------------------------------------------------------

export class Pruner {
  private tokenBudget: number;
  private minScoreThreshold: number;
  private alwaysKeepLastN: number;
  private summarizeBeforePrune: boolean;
  private scorer: MessageScorer;
  private summarizer: Summarizer;
  private preserveRoles: Set<string>;

  constructor(options: PrunerOptions = {}) {
    this.tokenBudget = options.tokenBudget ?? 4096;
    this.minScoreThreshold = options.minScoreThreshold ?? 0.3;
    this.alwaysKeepLastN = options.alwaysKeepLastN ?? 4;
    this.summarizeBeforePrune = options.summarizeBeforePrune ?? true;
    this.scorer = options.scorer ?? new MessageScorer();
    this.summarizer = options.summarizer ?? new Summarizer();
    this.preserveRoles = new Set(options.preserveRoles ?? ["system"]);
  }

  async prune(messages: Message[], force = false): Promise<PruneResult> {
    let msgs = [...messages];

    const currentTokens = () =>
      msgs.reduce((s, m) => s + estimateTokens(m.content), 0);

    if (!force && currentTokens() <= this.tokenBudget) {
      return {
        messages: msgs,
        removedCount: 0,
        removedTokens: 0,
        summaryInserted: false,
        scores: [],
      };
    }

    const [protected_, candidates] = this._partition(msgs);
    const scored = this.scorer.scoreMessages(msgs);
    const scoreByIdx = new Map(scored.map((s) => [s.index, s]));

    const candidateScored = candidates
      .map((i) => scoreByIdx.get(i)!)
      .filter(Boolean)
      .sort((a, b) => a.score - b.score);

    let removedCount = 0;
    let removedTokens = 0;
    let summaryInserted = false;
    let summaryResult: SummaryResult | undefined;

    // --- try summarization first ---
    if (this.summarizeBeforePrune && candidateScored.length) {
      const lowScored = candidateScored.filter(
        (s) => s.score < this.minScoreThreshold,
      );
      if (lowScored.length) {
        const chunk = lowScored.map((s) => msgs[s.index]);
        summaryResult = await this.summarizer.summarize(chunk);

        const saved =
          summaryResult.originalTokenCount - summaryResult.summaryTokenCount;
        if (saved > 0) {
          const summaryMsg = summaryAsMessage(summaryResult);
          const removeSet = new Set(lowScored.map((s) => s.index));
          const newMsgs: Message[] = [];
          let inserted = false;

          // Find the first non-protected, non-removed index for insertion
          const firstCandidate = Math.min(
            ...[...Array(msgs.length).keys()].filter(
              (i) => !protected_.has(i) && !removeSet.has(i),
            ),
          );

          for (let i = 0; i < msgs.length; i++) {
            if (removeSet.has(i)) {
              removedCount++;
              removedTokens += estimateTokens(msgs[i].content);
            } else {
              if (!inserted && i === firstCandidate) {
                newMsgs.push(summaryMsg);
                inserted = true;
              }
              newMsgs.push(msgs[i]);
            }
          }

          if (!inserted) {
            // insert after last system message
            let insertAt = 0;
            for (let i = 0; i < newMsgs.length; i++) {
              if (newMsgs[i].role === "system") insertAt = i + 1;
            }
            newMsgs.splice(insertAt, 0, summaryMsg);
          }

          msgs = newMsgs;
          summaryInserted = true;
        }
      }
    }

    // --- hard prune if still over budget ---
    if (currentTokens() > this.tokenBudget) {
      const scored2 = this.scorer.scoreMessages(msgs);
      const scoreByIdx2 = new Map(scored2.map((s) => [s.index, s]));
      const [, candidates2] = this._partition(msgs);
      const candsSort = candidates2
        .map((i) => scoreByIdx2.get(i)!)
        .filter(Boolean)
        .sort((a, b) => a.score - b.score);

      const removeSet2 = new Set<number>();
      let tokens = currentTokens();
      for (const sm of candsSort) {
        if (tokens <= this.tokenBudget) break;
        tokens -= sm.tokenCount;
        removeSet2.add(sm.index);
        removedCount++;
        removedTokens += sm.tokenCount;
      }

      msgs = msgs.filter((_, i) => !removeSet2.has(i));
    }

    return {
      messages: msgs,
      removedCount,
      removedTokens,
      summaryInserted,
      summaryResult,
      scores: scored,
    };
  }

  needsPruning(messages: Message[]): boolean {
    const total = messages.reduce((s, m) => s + estimateTokens(m.content), 0);
    return total > this.tokenBudget;
  }

  private _partition(msgs: Message[]): [Set<number>, number[]] {
    const n = msgs.length;
    const protected_ = new Set<number>();

    msgs.forEach((m, i) => {
      if (this.preserveRoles.has(m.role)) protected_.add(i);
    });

    const nonSystem = Array.from({ length: n }, (_, i) => i).filter(
      (i) => !protected_.has(i),
    );
    nonSystem.slice(-this.alwaysKeepLastN).forEach((i) => protected_.add(i));

    const candidates = Array.from({ length: n }, (_, i) => i).filter(
      (i) => !protected_.has(i),
    );
    return [protected_, candidates];
  }
}

File: neuromem/pruner.py
"""
neuromem.pruner
~~~~~~~~~~~~~~~
Safely prunes low-importance messages while respecting token budgets
and preserving critical content.

The pruner works in several passes:
  1. Always keep: system messages, the most recent N turns.
  2. Score remaining messages via :class:`~neuromem.scorer.MessageScorer`.
  3. Drop lowest-scored messages until the token budget is met.
  4. Optionally trigger summarization before hard pruning.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional, Sequence

from .scorer import MessageScorer, ScoredMessage, _estimate_tokens
from .summarizer import Summarizer, SummaryResult


# ---------------------------------------------------------------------------
# Result type
# ---------------------------------------------------------------------------

@dataclass
class PruneResult:
    """Outcome of a :meth:`Pruner.prune` operation."""
    messages: List[dict]              # final message list (ready to send)
    removed_count: int
    removed_tokens: int
    summary_inserted: bool
    summary_result: Optional[SummaryResult] = None
    scores: List[ScoredMessage] = field(default_factory=list)

    @property
    def total_tokens(self) -> int:
        return sum(_estimate_tokens(m.get("content", "")) for m in self.messages)

    def __repr__(self) -> str:  # pragma: no cover
        return (
            f"PruneResult(kept={len(self.messages)}, removed={self.removed_count}, "
            f"tokens={self.total_tokens}, summary={self.summary_inserted})"
        )


# ---------------------------------------------------------------------------
# Pruner
# ---------------------------------------------------------------------------

class Pruner:
    """
    Intelligently prunes a message list to fit within a token budget.

    Parameters
    ----------
    token_budget : int
        Hard upper-bound on total context tokens (default: 4096).
    min_score_threshold : float
        Messages below this score are candidates for pruning (0–1).
    always_keep_last_n : int
        Always preserve the most recent *n* turns regardless of score.
    summarize_before_prune : bool
        If True, try to summarize pruneable messages before hard-dropping them.
    scorer : MessageScorer, optional
        Custom scorer instance.
    summarizer : Summarizer, optional
        Custom summarizer instance.
    preserve_roles : list of str
        Roles that must never be pruned (default: ["system"]).
    """

    def __init__(
        self,
        token_budget: int = 4096,
        min_score_threshold: float = 0.3,
        always_keep_last_n: int = 4,
        summarize_before_prune: bool = True,
        scorer: Optional[MessageScorer] = None,
        summarizer: Optional[Summarizer] = None,
        preserve_roles: Optional[List[str]] = None,
    ) -> None:
        self.token_budget = token_budget
        self.min_score_threshold = min_score_threshold
        self.always_keep_last_n = always_keep_last_n
        self.summarize_before_prune = summarize_before_prune
        self.scorer = scorer or MessageScorer()
        self.summarizer = summarizer or Summarizer()
        self.preserve_roles = set(preserve_roles or ["system"])

    # ------------------------------------------------------------------

    def prune(
        self,
        messages: Sequence[dict],
        *,
        force: bool = False,
    ) -> PruneResult:
        """
        Prune *messages* to fit within :attr:`token_budget`.

        Parameters
        ----------
        messages : sequence of {"role": str, "content": str}
        force : bool
            If True, prune even if already under budget.

        Returns
        -------
        PruneResult
        """
        msgs = list(messages)

        current_tokens = sum(_estimate_tokens(m.get("content", "")) for m in msgs)
        if not force and current_tokens <= self.token_budget:
            return PruneResult(
                messages=msgs,
                removed_count=0,
                removed_tokens=0,
                summary_inserted=False,
            )

        # --- partition messages ---
        protected_indices, candidate_indices = self._partition(msgs)

        # Score candidate messages
        scored = self.scorer.score_messages(msgs)
        score_by_idx = {s.index: s for s in scored}

        candidate_scored = sorted(
            [score_by_idx[i] for i in candidate_indices if i in score_by_idx],
            key=lambda s: s.score,
        )

        removed_count = 0
        removed_tokens = 0
        summary_inserted = False
        summary_result: Optional[SummaryResult] = None

        # --- try summarization first ---
        if self.summarize_before_prune and candidate_scored:
            low_scored = [s for s in candidate_scored if s.score < self.min_score_threshold]
            if low_scored:
                chunk = [msgs[s.index] for s in low_scored]
                summary_result = self.summarizer.summarize(chunk)

                # Only insert summary if it actually saves tokens
                saved = summary_result.original_token_count - summary_result.summary_token_count
                if saved > 0:
                    summary_msg = summary_result.as_message()
                    # Remove summarized messages from msgs
                    remove_set = {s.index for s in low_scored}
                    new_msgs: List[dict] = []
                    inserted = False
                    for i, m in enumerate(msgs):
                        if i in remove_set:
                            removed_count += 1
                            removed_tokens += _estimate_tokens(m.get("content", ""))
                        else:
                            if not inserted and i not in protected_indices and i == min(
                                j for j in range(len(msgs)) if j not in remove_set and j not in protected_indices
                            ):
                                new_msgs.append(summary_msg)
                                inserted = True
                            new_msgs.append(m)
                    if not inserted:
                        # insert right after last system message
                        insert_at = 0
                        for i, m in enumerate(new_msgs):
                            if m.get("role") == "system":
                                insert_at = i + 1
                        new_msgs.insert(insert_at, summary_msg)

                    msgs = new_msgs
                    summary_inserted = True
                    current_tokens = sum(_estimate_tokens(m.get("content", "")) for m in msgs)

        # --- hard prune if still over budget ---
        if current_tokens > self.token_budget:
            # Re-score after potential summary insertion
            scored2 = self.scorer.score_messages(msgs)
            score_by_idx2 = {s.index: s for s in scored2}
            protected2, candidates2 = self._partition(msgs)
            cands_sorted = sorted(
                [score_by_idx2[i] for i in candidates2 if i in score_by_idx2],
                key=lambda s: s.score,
            )

            remove_set2: set[int] = set()
            for sm in cands_sorted:
                if current_tokens <= self.token_budget:
                    break
                current_tokens -= sm.token_count
                remove_set2.add(sm.index)
                removed_count += 1
                removed_tokens += sm.token_count

            msgs = [m for i, m in enumerate(msgs) if i not in remove_set2]

        return PruneResult(
            messages=msgs,
            removed_count=removed_count,
            removed_tokens=removed_tokens,
            summary_inserted=summary_inserted,
            summary_result=summary_result,
            scores=scored,
        )

    # ------------------------------------------------------------------

    def _partition(self, msgs: List[dict]) -> tuple[set[int], set[int]]:
        """Split indices into (protected, candidates)."""
        n = len(msgs)
        protected: set[int] = set()

        for i, m in enumerate(msgs):
            if m.get("role") in self.preserve_roles:
                protected.add(i)

        # Always-keep last N non-system messages
        non_system = [i for i in range(n) if i not in protected]
        for idx in non_system[-self.always_keep_last_n:]:
            protected.add(idx)

        candidates = set(range(n)) - protected
        return protected, candidates

    # ------------------------------------------------------------------

    def needs_pruning(self, messages: Sequence[dict]) -> bool:
        """Return True if messages exceed the token budget."""
        total = sum(_estimate_tokens(m.get("content", "")) for m in messages)
        return total > self.token_budget
